File: leetcode/x-of-a-kind-in-a-deck-of-cards.go
package main

func steinsGcd(a int, b int) int {
	if a == 0 {
		return b
	}

	if b == 0 {
		return a
	}

	// Find greatest power of 2 that
	// divides both a and b
	pow := 0
	for pow = 0; ((a | b) & 1) == 0; pow++ {
		a = a >> 1
		b = b >> 1
	}

	// Make sure a is odd
	for a&1 == 0 {
		a = a >> 1
	}

	for {
		// Make sure b is odd
		for b&1 == 0 {
			b = b >> 1
		}

		if a > b {
			a, b = b, a
		}
		b = b - a
		if b == 0 {
			return a << pow
		}
	}
}

func findGcd(a int, b int) int {
	for a != 0 {
		a, b = b%a, a
	}
	return b
}

func hasGroupsSizeX(deck []int) bool {

	if len(deck) < 2 {
		return false
	}

	freqMap := make(map[int]int)
	for _, card := range deck {
		freqMap[card]++
	}

	freqs := make([]int, len(freqMap))
	for _, freq := range freqMap {
		freqs = append(freqs, freq)
	}

	gcd := steinsGcd(freqs[0], freqs[1])
	for _, freq := range freqs {
		gcd = steinsGcd(gcd, freq)
	}

	if gcd < 2 {
		return false
	}

	return true
}

File: leetcode/x-of-a-kind-in-a-deck-of-cards.cpp
#include <vector>
#include <map>
using namespace std;
#define vi vector<int>
#define mii map<int,int>

class Solution
{

private:
    int steins_gcd(int a, int b) {
        if (a == 0) return b;
        if (b == 0) return a;

        // find highest power of 2 that divides both a and b
        int pow;
        for(pow=0; ((a|b)&1)==0; pow++) {
            a >>= 1;
            b >>= 1;
        }

        // make sure a is odd
        while((a&1) == 0) 
            a >>= 1;

        while (true) {
            while ((b&1) == 0) 
                b >>= 1;
            
            if (a > b) {
                int temp = b; b = a; a = temp;
            }

            b = b - a;

            if (b == 0)
                return a<<pow;
        }
    }

    int find_gcd(int a, int b) {
        int temp;
        while (a != 0) {
            temp = a;
            a = b % a;
            b = temp;
        }
        return b;
    }
public:
    bool hasGroupsSizeX(vi &deck)
    {
        mii freqMap = mii();
        for (const auto &card: deck)
            freqMap[card]++;

        vi freqs = vi(freqMap.size());
        for (const auto &map_value: freqMap)
            freqs.push_back(map_value.second);

        if (freqs.size() < 2)
            return false;

        int gcd = steins_gcd(freqs[0], freqs[1]);
        for (const auto &freq: freqs)
            gcd = steins_gcd(gcd, freq);

        if (gcd < 2)
            return false;
        
        return true;
    }
};

int main() {

}